File: plugins/keyboard/ios/templates/EmojiSearchIndex.swift
import Foundation

// Keyword sources: (1) Unicode scalar names via Unicode.Scalar.Properties.name
// — exhaustive and locale-stable; (2) `manualKeywords` for colloquial terms
// the Unicode names miss ("happy" for 😀, "love" for ❤️). Substring scan over
// a few hundred entries; no trie / scoring crate.
struct EmojiSearchIndex {

    private struct Entry {
        let emoji: String
        let keywords: [String]
    }

    private let entries: [Entry]

    init() {
        var built: [Entry] = []
        var seen = Set<String>()
        for cat in EmojiCategory.allCases where cat != .recents {
            for emoji in EmojiData.emojis(for: cat) {
                guard seen.insert(emoji).inserted else { continue }
                let unicode = EmojiSearchIndex.unicodeKeywords(for: emoji)
                let manual = EmojiSearchIndex.manualKeywords[emoji] ?? []
                let combined = Array(Set(unicode + manual))
                built.append(Entry(emoji: emoji, keywords: combined))
            }
        }
        self.entries = built
    }

    func search(_ query: String) -> [String] {
        let q = query
            .lowercased()
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !q.isEmpty else { return [] }
        var matches: [String] = []
        matches.reserveCapacity(min(entries.count, 60))
        for entry in entries {
            for kw in entry.keywords where kw.contains(q) {
                matches.append(entry.emoji)
                break
            }
        }
        return matches
    }

    // Variation selectors / ZWJ have no scalar name and are skipped silently.
    private static func unicodeKeywords(for emoji: String) -> [String] {
        var words = Set<String>()
        for scalar in emoji.unicodeScalars {
            guard let name = scalar.properties.name else { continue }
            for word in name.lowercased().split(separator: " ") {
                words.insert(String(word))
            }
        }
        return Array(words)
    }

    private static let manualKeywords: [String: [String]] = [
        "😀": ["happy", "smile", "grin", "joy"],
        "😃": ["happy", "smile", "open"],
        "😄": ["happy", "smile", "joy"],
        "😁": ["happy", "smile", "beam", "teeth"],
        "😆": ["laugh", "smile", "happy", "haha"],
        "😅": ["sweat", "smile", "phew", "relief"],
        "🤣": ["laugh", "rofl", "rolling", "haha"],
        "😂": ["laugh", "tears", "joy", "haha", "cry"],
        "🙂": ["smile", "slight"],
        "🙃": ["upside", "down", "silly"],
        "😉": ["wink"],
        "😊": ["happy", "smile", "blush"],
        "😍": ["love", "heart", "eyes"],
        "🥰": ["love", "hearts", "happy"],
        "😘": ["kiss", "love"],
        "😎": ["cool", "sunglasses"],
        "🤔": ["think", "hmm", "thinking"],
        "😢": ["cry", "sad", "tear"],
        "😭": ["cry", "sob", "sad"],
        "😡": ["angry", "mad", "rage"],
        "🥺": ["plead", "puppy", "begging"],
        "😴": ["sleep", "zzz", "tired"],
        "😱": ["scream", "shock", "fear"],
        "😮": ["wow", "shock", "surprise"],
        "🤯": ["mind", "blown", "wow"],
        "👍": ["thumbs", "up", "yes", "like", "ok"],
        "👎": ["thumbs", "down", "no", "dislike"],
        "👏": ["clap", "applause"],
        "🙏": ["pray", "thanks", "please"],
        "❤️": ["love", "heart", "red"],
        "💔": ["heartbreak", "sad", "broken"],
        "💯": ["100", "hundred", "perfect"],
        "🔥": ["fire", "lit", "hot"],
        "✨": ["sparkle", "shine", "magic"],
        "🐶": ["dog", "puppy"],
        "🐱": ["cat", "kitten"],
        "🦊": ["fox"],
        "🐼": ["panda"],
        "🦁": ["lion"],
        "🐸": ["frog"],
        "🍕": ["pizza"],
        "🍔": ["burger", "hamburger"],
        "🍟": ["fries", "chips"],
        "🌮": ["taco"],
        "🍣": ["sushi"],
        "☕️": ["coffee", "drink"],
        "🍺": ["beer"],
        "🎉": ["party", "celebrate"],
        "🎂": ["birthday", "cake"],
        "💩": ["poop", "shit"],
        "👀": ["eyes", "look"],
    ]
}

File: plugins/keyboard/android/templates/EmojiSearchIndex.kt
package com.a1lab.echos.ime

import android.content.Context

// Keyword sources: (1) Character.getName for Unicode scalar names; (2)
// `manualKeywords` for colloquial terms ("happy" for 😀, "love" for ❤️).
// Substring scan over a few hundred entries; no trie / scoring library.
class EmojiSearchIndex(context: Context) {

    private data class Entry(
        val emoji: String,
        val keywords: List<String>,
    )

    private val entries: List<Entry>

    init {
        val seen = HashSet<String>()
        val built = mutableListOf<Entry>()
        for (category in EmojiCategory.values()) {
            if (category == EmojiCategory.RECENTS) continue
            for (emoji in EmojiData.emojis(category, context)) {
                if (!seen.add(emoji)) continue
                val unicode = unicodeKeywords(emoji)
                val manual = manualKeywords[emoji].orEmpty()
                built.add(Entry(emoji, (unicode + manual).distinct()))
            }
        }
        entries = built
    }

    fun search(query: String): List<String> {
        val q = query.lowercase().trim()
        if (q.isEmpty()) return emptyList()
        val matches = ArrayList<String>(minOf(entries.size, 60))
        for (entry in entries) {
            for (kw in entry.keywords) {
                if (kw.contains(q)) {
                    matches.add(entry.emoji)
                    break
                }
            }
        }
        return matches
    }

    // Variation selectors / ZWJ have no codepoint name and are skipped silently.
    private fun unicodeKeywords(emoji: String): List<String> {
        val words = HashSet<String>()
        var i = 0
        while (i < emoji.length) {
            val cp = emoji.codePointAt(i)
            val name = try { Character.getName(cp) } catch (_: IllegalArgumentException) { null }
            if (name != null) {
                for (word in name.lowercase().split(' ')) {
                    if (word.isNotEmpty()) words.add(word)
                }
            }
            i += Character.charCount(cp)
        }
        return words.toList()
    }

    companion object {
        private val manualKeywords: Map<String, List<String>> = mapOf(
            "😀" to listOf("happy", "smile", "grin", "joy"),
            "😃" to listOf("happy", "smile", "open"),
            "😄" to listOf("happy", "smile", "joy"),
            "😁" to listOf("happy", "smile", "beam", "teeth"),
            "😆" to listOf("laugh", "smile", "happy", "haha"),
            "😅" to listOf("sweat", "smile", "phew", "relief"),
            "🤣" to listOf("laugh", "rofl", "rolling", "haha"),
            "😂" to listOf("laugh", "tears", "joy", "haha", "cry"),
            "🙂" to listOf("smile", "slight"),
            "🙃" to listOf("upside", "down", "silly"),
            "😉" to listOf("wink"),
            "😊" to listOf("happy", "smile", "blush"),
            "😍" to listOf("love", "heart", "eyes"),
            "🥰" to listOf("love", "hearts", "happy"),
            "😘" to listOf("kiss", "love"),
            "😎" to listOf("cool", "sunglasses"),
            "🤔" to listOf("think", "hmm", "thinking"),
            "😢" to listOf("cry", "sad", "tear"),
            "😭" to listOf("cry", "sob", "sad"),
            "😡" to listOf("angry", "mad", "rage"),
            "🥺" to listOf("plead", "puppy", "begging"),
            "😴" to listOf("sleep", "zzz", "tired"),
            "😱" to listOf("scream", "shock", "fear"),
            "😮" to listOf("wow", "shock", "surprise"),
            "🤯" to listOf("mind", "blown", "wow"),
            "👍" to listOf("thumbs", "up", "yes", "like", "ok"),
            "👎" to listOf("thumbs", "down", "no", "dislike"),
            "👏" to listOf("clap", "applause"),
            "🙏" to listOf("pray", "thanks", "please"),
            "❤️" to listOf("love", "heart", "red"),
            "💔" to listOf("heartbreak", "sad", "broken"),
            "💯" to listOf("100", "hundred", "perfect"),
            "🔥" to listOf("fire", "lit", "hot"),
            "✨" to listOf("sparkle", "shine", "magic"),
            "🐶" to listOf("dog", "puppy"),
            "🐱" to listOf("cat", "kitten"),
            "🦊" to listOf("fox"),
            "🐼" to listOf("panda"),
            "🦁" to listOf("lion"),
            "🐸" to listOf("frog"),
            "🍕" to listOf("pizza"),
            "🍔" to listOf("burger", "hamburger"),
            "🍟" to listOf("fries", "chips"),
            "🌮" to listOf("taco"),
            "🍣" to listOf("sushi"),
            "☕️" to listOf("coffee", "drink"),
            "🍺" to listOf("beer"),
            "🎉" to listOf("party", "celebrate"),
            "🎂" to listOf("birthday", "cake"),
            "💩" to listOf("poop", "shit"),
            "👀" to listOf("eyes", "look"),
        )
    }
}
